Allow terraform-bundle to include custom plugins

ScottWinkler opened the pull request:
This commit allows for terraform-bundle to be able to package custom private plugins and addresses support ticket #7214. Now, terraform-bundle will attempt to look for a plugin on the public registry, and if that fails, it will look in a local plugins directory. If the plugin can be found in the local folder, that plugin will be included in the bundle. By default the local plugins directory is "./plugins", but this can be changed by setting the -plugins-dir argument. Ex: 

my test.hcl file with a custom plugin that does not exist on the public registry
```
providers {
  # Include the newest "aws" provider version in the 1.0 series.
  aws = ["~> 1.0"]

  # Include both the newest 1.0 and 2.0 versions of the "google" provider.
  # Each item in these lists allows a distinct version to be added. If the
  # two expressions match different versions then _both_ are included in
  # the bundle archive.
  google = ["~> 1.0"]

 mycustomplugin = ["0.1"] # lives at ./my-plugins/terraform-provider-mycustomplugin
}
```

use case
```
$ terraform-bundle -os=linux -arch=amd64 -plugin-dir=./my-plugins test.hcl
```

apparentlymart commented:
Thanks for working on this, @ScottWinkler! This seems like a good idea to me, to make it easier to combine both first- and third-party plugins together into a single bundle.

From reading the code here it looks like as currently implemented it uses the plugin directory as a fallback for the "official" repository, only looking there if an official release does not exist. What do you think about inverting that so that the local directory is used in preference? That would allow also including local custom builds of providers that would normally be installed from the main server, and would behave consistently with how Terraform's own installer works where already-installed local plugins take precedence over remote ones.

We have a utility function [FindPlugins](https://godoc.org/github.com/hashicorp/terraform/plugin/discovery#FindPlugins) which can be used to find plugins in a particular directory, and then the `PluginMetaSet` methods can be used to see if there are locally-installed plugins matching the given constraints, and then produce a set of plugins that need to be installed from the remote server. Using these functions would ensure the behavior is consistent with the `terraform init` behavior.

I'd love to hear if you think the above would be reasonable adjustments to what you proposed here. If you agree but don't have the time to do another round of changes then please do let me know, in which case we'd be happy to make such adjustments ourselves to get this in.

ScottWinkler commented:
Good suggestion @apparentlymart! I updated the code to use the FindPlugins utility function and tested it with my 3rd party plugin, and it works great. I also changed the ordering, to look first in the local directory, then to look at the public registry

apparentlymart commented:
Thanks for updating this, @ScottWinkler! I've merged this change.

Since this tool isn't part of the Terraform CLI itself, those using this tool will need to build a new version locally from the latest source in order to get this additional functionality.
